From f548b3e2abaabefffdb8991694f9dffa3fb31836 Mon Sep 17 00:00:00 2001
From: Yuri Tolstov <yuri.tolstov@cavium.com>
Date: Fri, 19 Oct 2018 10:04:35 +0300
Subject: [PATCH 0328/1051] octeontx-pko: Adjust PKO FC settings to number of
 buffers

PKO DQ flow-control settings are aligned with the number of PKO
buffers used per DQ. The maximum number of DQ buffers should not be
bigger than total number of buffers used by PKO. The flow-control is
setup as a half of the buffers used by DQ.

Signed-off-by: Yuri Tolstov <yuri.tolstov@cavium.com>
Signed-off-by: Yury Norov <ynorov@marvell.com>
Signed-off-by: Yury Norov <ynorov@caviumnetworks.com>
[Kevin: The original patch got from Marvell sdk10.0_19.06]
Signed-off-by: Kevin Hao <kexin.hao@windriver.com>
---
 .../cavium/octeontx-83xx/pkopf_main.c         | 27 ++++++++++++-------
 1 file changed, 17 insertions(+), 10 deletions(-)

diff --git a/drivers/net/ethernet/cavium/octeontx-83xx/pkopf_main.c b/drivers/net/ethernet/cavium/octeontx-83xx/pkopf_main.c
index 8d489d7c9e8a..d5ec8641c7a6 100644
--- a/drivers/net/ethernet/cavium/octeontx-83xx/pkopf_main.c
+++ b/drivers/net/ethernet/cavium/octeontx-83xx/pkopf_main.c
@@ -30,6 +30,16 @@
 #define BGX_CHAN_BASE	0x800
 #define BGX_CHAN_RANGE	BIT(8)
 
+#define PKO_BUFFERS	4096
+#define PKO_MAX_PORTS	16 /* Maximum number simultaneously used ports.*/
+#define PKO_MAX_DQ_BUFS	(PKO_BUFFERS / PKO_MAX_PORTS)
+
+/* SKID value should be synchronized with DQ setting in RM user program.
+ * Each DQ has maxiumum depth of PKO_MAX_DQ_BUFS FPA 4KB buffers.
+ * SKID is half of this value.
+ */
+#define PKO_FC_SKID	(PKO_MAX_DQ_BUFS / 2)
+
 static atomic_t pko_count = ATOMIC_INIT(0);
 static DEFINE_MUTEX(octeontx_pko_devices_lock);
 static LIST_HEAD(octeontx_pko_devices);
@@ -1088,9 +1098,7 @@ static int pko_disable(struct pkopf *pko)
 
 static int setup_dpfi(struct pkopf *pko)
 {
-	int err;
-	int buffers;
-	int retry = 0;
+	int err, retry = 0;
 	u64 reg;
 
 	err = fpapf->create_domain(pko->id, FPA_PKO_DPFI_GMID, 1, NULL);
@@ -1107,10 +1115,8 @@ static int setup_dpfi(struct pkopf *pko)
 		symbol_put(fpavf_com);
 		return -ENODEV;
 	}
-	buffers = 4096;
-
-	err = fpavf->setup(fpa, buffers, pko->pdm_buf_size,
-			FPA_VF_FLAG_CONT_MEM);
+	err = fpavf->setup(fpa, PKO_BUFFERS, pko->pdm_buf_size,
+			   FPA_VF_FLAG_CONT_MEM);
 	if (err) {
 		dev_err(&pko->pdev->dev, "failed to setup fpavf\n");
 		symbol_put(fpapf_com);
@@ -1122,6 +1128,7 @@ static int setup_dpfi(struct pkopf *pko)
 	pko_reg_write(pko, PKO_PF_DPFI_GMCTL, FPA_PKO_DPFI_GMID);
 	pko_reg_write(pko, PKO_PF_DPFI_FLUSH, 0);
 	pko_reg_write(pko, PKO_PF_DPFI_ENA, 0x1);
+
 	while (true) {
 		reg = pko_reg_read(pko, PKO_PF_DPFI_STATUS);
 		if (!(reg & 0x2))
@@ -1169,7 +1176,7 @@ static int pko_init(struct pkopf *pko)
 {
 	u64 reg;
 	int retry = 0;
-	int n = 1023;
+	int n;
 	int i;
 
 	reg = pko_reg_read(pko, PKO_PF_CONST);
@@ -1190,9 +1197,9 @@ static int pko_init(struct pkopf *pko)
 			return -ENODEV;
 	}
 
-	reg = 0;
 	reg = PKO_PDM_CFG_SET_PAD_MINLEN(PKO_PAD_MINLEN) |
-		PKO_PDM_CFG_SET_DQ_FC_SKID(n) | PKO_PDM_CFG_SET_EN(1);
+		PKO_PDM_CFG_SET_DQ_FC_SKID(PKO_FC_SKID) |
+		PKO_PDM_CFG_SET_EN(1);
 	pko_reg_write(pko, PKO_PF_PDM_CFG, reg);
 
 	pko_reg_write(pko, PKO_PF_SHAPER_CFG, 0x1);
-- 
2.17.1

